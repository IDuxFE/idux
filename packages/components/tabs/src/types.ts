/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { DefineComponent, FunctionalComponent, HTMLAttributes, PropType, VNodeChild } from 'vue'

import { SelectData } from '@idux/components/select'

export type TabsMode = 'default' | 'primary'
export type TabsPlacement = 'top' | 'bottom' | 'start' | 'end'
export type TabsSize = 'lg' | 'md'
export type TabsType = 'card' | 'line' | 'segment'

export const tabsProps = {
  selectedKey: { type: [Number, String, Symbol] as PropType<VKey>, default: undefined },

  addable: { type: Boolean, default: false },
  closable: { type: Boolean, default: false },
  dataSource: { type: Array as PropType<TabsData[]>, default: undefined },
  forceRender: { type: Boolean, default: false },
  mode: { type: String as PropType<TabsMode>, default: 'default' },
  placement: { type: String as PropType<TabsPlacement>, default: 'top' },
  size: { type: String as PropType<TabsSize>, default: undefined },
  type: { type: String as PropType<TabsType>, default: 'card' },

  'onUpdate:selectedKey': [Function, Array] as PropType<MaybeArray<(key: any) => void>>,
  onAdd: [Function, Array] as PropType<MaybeArray<() => void | boolean | Promise<boolean>>>,
  onClose: [Function, Array] as PropType<MaybeArray<(key: any) => void | boolean | Promise<boolean>>>,
  /**
   * @deprecated
   */
  onTabClick: [Function, Array] as PropType<MaybeArray<(key: any, evt: Event) => void>>,
  onBeforeLeave: [Function, Array] as PropType<
    MaybeArray<(key: any, oldKey?: any) => void | boolean | Promise<boolean>>
  >,
} as const

export type TabsProps = ExtractInnerPropTypes<typeof tabsProps>
export type TabsPublicProps = Omit<ExtractPublicPropTypes<typeof tabsProps>, 'onTabClick'>
export type TabsComponent = DefineComponent<Omit<HTMLAttributes, keyof TabsPublicProps> & TabsPublicProps>
export type TabsInstance = InstanceType<DefineComponent<TabsProps>>

export interface TabProps {
  closable?: boolean
  content?: string
  disabled?: boolean
  forceRender?: boolean
  title?: string
}
export type TabComponent = FunctionalComponent<Omit<HTMLAttributes, keyof TabProps> & TabProps>

export interface TabsData<K = VKey> extends TabProps {
  key: K
  customContent?: string | ((data: { key: VKey; content?: string; selected?: boolean }) => VNodeChild)
  customTitle?: string | ((data: { key: VKey; disabled?: boolean; selected?: boolean; title?: string }) => VNodeChild)
  [key: string]: any
}

// private
export const tabNavProps = {
  closable: { type: Boolean, default: undefined },
  disabled: { type: Boolean, default: undefined },
  selected: { type: Boolean, default: undefined },
  title: { type: String, default: undefined },
  onSelected: { type: Function, required: true },
} as const

export const tabPaneProps = {
  content: { type: String, default: undefined },
  forceRender: { type: Boolean, default: undefined },
  selected: { type: Boolean, default: undefined },
} as const

export const moreSelectPaneProps = {
  dataSource: { type: Array as PropType<SelectData[]>, default: () => [] },
  visible: { type: Boolean, default: false },

  // private
  _virtualScrollHeight: { type: Number, default: 186 },
} as const
