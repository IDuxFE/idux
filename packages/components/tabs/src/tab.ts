/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import type { TabComponent } from './types'

const tabKey = Symbol('IxTab')
const Tab = (() => {}) as TabComponent
Tab.displayName = 'IxTab'
;(Tab as any)[tabKey] = true

export { Tab, tabKey }
