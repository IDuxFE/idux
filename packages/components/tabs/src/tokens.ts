/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { TabsData, TabsProps } from './types'
import type { VKey } from '@idux/cdk/utils'
import type { ComputedRef, InjectionKey } from 'vue'

export interface TabsContext {
  props: TabsProps
  mergedPrefixCls: ComputedRef<string>
  mergedDataSource: ComputedRef<TabsData[]>
  isHorizontal: ComputedRef<boolean>
  closedKeys: ComputedRef<VKey[]>
  navAttrMap: Map<VKey, { offset: number; size: number }>
  handleTabClick: (key: VKey, evt: Event) => Promise<void>
  handleTabClose: (key: VKey) => Promise<void>
  setSelectedKey: (value: VKey) => void
}

export const tabsToken: InjectionKey<TabsContext> = Symbol('tabsToken')
