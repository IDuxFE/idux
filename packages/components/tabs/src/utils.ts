/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

export function getMarginSize(el: Element, isHorizontal: boolean): number {
  const style = window.getComputedStyle(el)
  if (isHorizontal) {
    const marginLeft = parseFloat(style.marginLeft) || 0
    const marginRight = parseFloat(style.marginRight) || 0
    return marginLeft + marginRight
  } else {
    const marginTop = parseFloat(style.marginTop) || 0
    const marginBottom = parseFloat(style.marginBottom) || 0
    return marginTop + marginBottom
  }
}
