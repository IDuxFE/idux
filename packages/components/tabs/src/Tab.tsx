/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { defineComponent, inject } from 'vue'

import { tabsToken } from './tokens'
import { tabProps } from './types'

export default defineComponent({
  __IDUX_TAB: true,
  name: 'IxTab',
  props: tabProps,
  setup(props, { slots }) {
    const { mergedPrefixCls } = inject(tabsToken)!

    return () => {
      return <div class={`${mergedPrefixCls.value}-pane`}>{slots.default?.()}</div>
    }
  },
})
