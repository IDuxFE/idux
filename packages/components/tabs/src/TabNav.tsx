/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, inject, normalizeClass, ref, watchEffect } from 'vue'

import { useKey } from '@idux/components/utils'

import { tabsToken } from './tokens'
import { tabNavProps } from './types'

export default defineComponent({
  name: 'IxTabNav',
  props: tabNavProps,
  setup(props, { slots }) {
    const key = useKey()

    const { selectedKey, selectedElRef, mergedPrefixCls, handleTabClick } = inject(tabsToken)!

    const selfElRef = ref<HTMLElement | null>(null)
    const isSelected = computed(() => (selectedKey.value ?? props.defaultSelectedKey) === key)
    const prefixCls = computed(() => `${mergedPrefixCls.value}-nav`)
    const classes = computed(() => {
      return normalizeClass({
        [`${prefixCls.value}-tab`]: true,
        [`${prefixCls.value}-tab-selected`]: isSelected.value,
        [`${prefixCls.value}-tab-disabled`]: props.disabled,
      })
    })

    watchEffect(() => {
      if (isSelected.value && selfElRef.value) {
        selectedElRef.value = selfElRef.value
      }
    })

    const onClick = (evt: Event) => {
      if (!props.disabled) {
        handleTabClick(key, evt)
      }
    }

    return () => {
      const tab = <span class={`${prefixCls.value}-tab-label`}> {slots.title?.() ?? props.title}</span>
      return (
        <div class={classes.value} onClick={onClick} ref={selfElRef}>
          {tab}
        </div>
      )
    }
  },
})
