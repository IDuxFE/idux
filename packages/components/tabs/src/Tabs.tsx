/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, normalizeClass, provide, reactive } from 'vue'

import { isNil, isString } from 'lodash-es'

import { type VKey, callEmit, useControlledProp, useState } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'
import { useThemeToken } from '@idux/components/theme'

import { useDataSource } from './composables/useDataSource'
import TabNavWrapper from './contents/TabNavWrapper'
import TabPane from './contents/TabPane'
import { tabsToken } from './tokens'
import { type TabsData, tabsProps } from './types'
import { getThemeTokens } from '../theme'

export default defineComponent({
  name: 'IxTabs',
  props: tabsProps,
  setup(props, { slots }) {
    const common = useGlobalConfig('common')
    const { globalHashId, hashId, registerToken } = useThemeToken('tabs')
    registerToken(getThemeTokens)

    const config = useGlobalConfig('tabs')

    const mergedPrefixCls = computed(() => `${common.prefixCls}-tabs`)
    const mergedSize = computed(() => props.size ?? config.size)

    const mergedDataSource = useDataSource(props, slots)
    const horizontalPlacement = ['top', 'bottom']
    const isHorizontal = computed(() => horizontalPlacement.includes(props.placement))

    const [selectedKey, setSelectedKey] = useControlledProp(props, 'selectedKey')
    const [closedKeys, setClosedKeys] = useState<VKey[]>([])

    // 存储每个标签的尺寸和偏移
    const navAttrs = reactive<Record<VKey, { offset: number; size: number }>>({})

    const handleTabClick = async (key: VKey, evt: Event) => {
      const result = await callEmit(props.onBeforeLeave, key, selectedKey.value)
      if (result !== false) {
        setSelectedKey(key)
        /**
         * @deprecated
         */
        callEmit(props.onTabClick, key, evt)
      }
    }

    const handleTabClose = async (key: VKey) => {
      const result = await callEmit(props.onClose, key)
      if (result !== false) {
        const currSelectedKey = selectedKey.value
        const currClosedKeys = closedKeys.value

        if (key === selectedKey.value) {
          setSelectedKey(getNextSelectedKey(mergedDataSource.value, currClosedKeys, currSelectedKey))
        }

        setClosedKeys([...currClosedKeys, key])
      }
    }

    provide(tabsToken, {
      props,
      mergedPrefixCls,
      mergedDataSource,
      isHorizontal,
      closedKeys,
      navAttrs,
      handleTabClick,
      handleTabClose,
      setSelectedKey,
    })

    const classes = computed(() => {
      const { type, placement, mode } = props
      const prefixCls = mergedPrefixCls.value
      const size = mergedSize.value
      return normalizeClass({
        [globalHashId.value]: !!globalHashId.value,
        [hashId.value]: !!hashId.value,
        [prefixCls]: true,
        [`${prefixCls}-${size}`]: true,
        [`${prefixCls}-${type}`]: true,
        [`${prefixCls}-nav-${placement}`]: placement === 'top' || type === 'line',
        [`${prefixCls}-nav-${mode}`]: type === 'segment',
      })
    })

    return () => {
      const dataSource = mergedDataSource.value
      const currClosedKeys = closedKeys.value
      const currSelectedKey = selectedKey.value ?? getNextSelectedKey(dataSource, currClosedKeys)
      return (
        <div class={classes.value}>
          <TabNavWrapper selectedKey={currSelectedKey} v-slots={slots} />
          <div class={`${mergedPrefixCls.value}-pane-wrapper`}>
            {dataSource.map(data => {
              const { key, content, forceRender, customContent = 'content' } = data
              const contentSlot = isString(customContent) ? slots[customContent] : customContent
              return (
                <TabPane
                  key={key}
                  closed={currClosedKeys.includes(key)}
                  content={content}
                  forceRender={forceRender}
                  selected={key === currSelectedKey}
                  v-slots={{ content: contentSlot }}
                />
              )
            })}
          </div>
        </div>
      )
    }
  },
})

function getNextSelectedKey(dataSource: TabsData[], closedKeys: VKey[], currSelectedKey?: VKey) {
  const isValidNext = (data: TabsData) => !data.disabled && !closedKeys.includes(data.key)

  const currSelectedIndex = isNil(currSelectedKey) ? -1 : dataSource.findIndex(item => item.key === currSelectedKey)

  if (currSelectedIndex === -1) {
    return dataSource.find(isValidNext)?.key
  }

  for (let index = currSelectedIndex + 1; index < dataSource.length; index++) {
    const data = dataSource[index]
    if (isValidNext(data)) {
      return data.key
    }
  }

  for (let index = currSelectedIndex - 1; index > 0; index--) {
    const data = dataSource[index]
    if (isValidNext(data)) {
      return data.key
    }
  }

  return
}
