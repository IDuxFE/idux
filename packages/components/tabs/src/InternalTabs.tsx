/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { TabProps, TabsProps } from './types'
import type { VKey } from '@idux/cdk/utils'
import type { IconInstance } from '@idux/components/icon'
import type { CSSProperties, ComputedRef, PropType, Ref, VNode } from 'vue'

import { computed, defineComponent, normalizeClass, provide, ref, vShow, watch, withDirectives } from 'vue'

import { curry, isNil } from 'lodash-es'

import { useResizeObserver } from '@idux/cdk/resize'
import { addClass, callEmit, removeClass, useControlledProp, useState } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'
import { IxIcon } from '@idux/components/icon'

import TabNav from './TabNav'
import { useSelectedElOffset } from './composables/useOffset'
import { useNavRelatedElSize, useSelectedElVisibleSize } from './composables/useSize'
import { tabsToken } from './tokens'
import { tabsProps } from './types'

function useNavPreNextClasses(
  props: TabsProps,
  mergedPrefixCls: ComputedRef<string>,
  type: 'pre' | 'next',
  disabled: Ref<boolean>,
) {
  return computed(() => {
    const { placement } = props
    const prefixCls = mergedPrefixCls.value
    return normalizeClass({
      [`${prefixCls}-nav-${type}`]: true,
      [`${prefixCls}-nav-${type}-disabled`]: disabled.value,
      [`${prefixCls}-nav-${type}-${placement}`]: true,
    })
  })
}

function filterTabVNodes(
  props: TabsProps,
  tabVNodes: VNode[],
  selectedKey: ComputedRef<VKey | undefined>,
  defaultSelectedKey: VKey | undefined,
): VNode[] {
  const renderTabVNodes: VNode[] = []
  tabVNodes.forEach(vNode => {
    const { key } = vNode
    const { forceRender, disabled } = vNode.props as TabProps
    const _disabled = !isNil(disabled)
    const useVShow = forceRender ?? props.forceRender
    const show = (selectedKey.value ?? defaultSelectedKey) === key && !_disabled
    if (useVShow) {
      renderTabVNodes.push(withDirectives(vNode, [[vShow, show]]))
    } else if (show) {
      renderTabVNodes.push(vNode)
    }
  })
  return renderTabVNodes
}

export default defineComponent({
  props: {
    ...tabsProps,
    tabs: { type: Array as PropType<VNode[]>, default: undefined },
  },
  setup(props) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-tabs`)

    const navWrapperElRef = ref<HTMLElement | null>(null)
    const navElRef = ref<HTMLElement | null>(null)
    const navBarElRef = ref<HTMLElement | null>(null)
    const navPreElRef = ref<IconInstance | null>(null)
    const selectedElRef = ref<HTMLElement | null>(null)

    const [selectedKey, setSelectedKey] = useControlledProp(props, 'selectedKey')

    const isLineType = computed(() => props.type === 'line')
    const isSegmentType = computed(() => props.type === 'segment')

    const horizontalPlacement = ['top', 'bottom']
    const isHorizontal = computed(() => horizontalPlacement.includes(props.placement))

    const [navOffset, setNavOffset] = useState(0)
    const [barStyle, setBarStyle] = useState<CSSProperties>({})
    const {
      navSize,
      navWrapperSize,
      navPreNextSize,
      selectedElSize,
      setNavElSize,
      setSelectedElSize,
      setNavPreNextElSize,
    } = useNavRelatedElSize(isHorizontal, navWrapperElRef, navElRef, navPreElRef, selectedElRef)
    const { selectedElOffset, setSelectedElOffset } = useSelectedElOffset(isHorizontal, navPreNextSize, selectedElRef)

    const hasScroll = computed(() => {
      return navSize.value > navWrapperSize.value
    })

    watch(
      hasScroll,
      () => {
        setNavPreNextElSize()
        updateNavBarStyle()
        updateSelectedOffset()
      },
      {
        flush: 'post',
      },
    )

    const selectedElVisibleSize = useSelectedElVisibleSize(navWrapperSize, selectedElOffset, navOffset)

    // 处理存在滚动状态下，滚动到被选中的tab，并修正其位置
    const updateSelectedOffset = () => {
      if (hasScroll.value) {
        const size = selectedElVisibleSize.value / navWrapperSize.value
        const inVisibleRange = size < 2
        if (inVisibleRange) {
          // 可视范围内需要处理展示不全的问题，需要修正
          if (selectedElVisibleSize.value < selectedElSize.value) {
            // 即可视范围内最后一个tab没有展示完全
            setNavOffset(navOffset.value + selectedElSize.value - selectedElVisibleSize.value + navPreNextSize.value)
          } else if (selectedElVisibleSize.value / navWrapperSize.value > 1) {
            // 即可视范围内第一个tab没有展示完全
            setNavOffset(
              navOffset.value - ((selectedElVisibleSize.value % navWrapperSize.value) + navPreNextSize.value),
            )
          }
        } else {
          setNavOffset(selectedElOffset.value - navPreNextSize.value)
        }
      }
    }

    const preReached = ref(false)
    const nextReached = ref(false)

    const classes = computed(() => {
      const { type, placement, mode } = props
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [prefixCls]: true,
        [`${prefixCls}-${type}`]: true,
        [`${prefixCls}-nav-${placement}`]: placement === 'top' || type === 'line',
        [`${prefixCls}-nav-${mode}`]: type === 'segment',
      })
    })

    const navWrapperClass = computed(() => {
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [`${prefixCls}-nav-wrapper`]: true,
        [`${prefixCls}-nav-wrapper-has-scroll`]: hasScroll.value,
      })
    })

    const curryNavPreNextClasses = curry(useNavPreNextClasses)(props, mergedPrefixCls)
    const navPreClasses = curryNavPreNextClasses('pre', preReached)
    const navNextClasses = curryNavPreNextClasses('next', nextReached)

    const handleTabClick = async (key: VKey, evt: Event) => {
      const leaveResult = callEmit(props.onBeforeLeave, key, selectedKey.value)
      const result = await leaveResult
      if (result !== false) {
        callEmit(props.onTabClick, key, evt)
        // 处理当前被选中元素再次被点击，需要修正其位置
        if (key === selectedKey.value) {
          updateSelectedOffset()
        }
        setSelectedKey(key)
      }
    }

    const updateNavBarStyle = () => {
      if (isLineType.value && navBarElRef.value) {
        const isBarDisabled = selectedElRef.value?.classList.contains(`${mergedPrefixCls.value}-nav-tab-disabled`)
        const barDisabledClassName = `${mergedPrefixCls.value}-nav-bar-disabled`
        const barOffset = selectedElOffset.value - navOffset.value + 'px'
        const barSize = selectedElSize.value + 'px'

        setBarStyle({
          width: isHorizontal.value ? barSize : '',
          left: isHorizontal.value ? barOffset : '',
          top: isHorizontal.value ? '' : barOffset,
          height: isHorizontal.value ? '' : barSize,
        })
        if (isBarDisabled) {
          addClass(navBarElRef.value, barDisabledClassName)
        } else {
          removeClass(navBarElRef.value, barDisabledClassName)
        }
      }
    }

    const handlePreClick = (evt: Event) => {
      if (!preReached.value) {
        callEmit(props.onPreClick, evt)
        const mergedOffset = navOffset.value + navPreNextSize.value
        const offset = mergedOffset < navWrapperSize.value ? 0 : mergedOffset - navWrapperSize.value
        setNavOffset(offset)
      }
    }

    const handleNextClick = (evt: Event) => {
      if (!nextReached.value) {
        callEmit(props.onNextClick, evt)
        const mergedNavSize = navSize.value! + navPreNextSize.value * 2
        const _offset = navOffset.value + navWrapperSize.value
        let offset
        if (mergedNavSize - _offset < navWrapperSize.value) {
          offset = mergedNavSize - navWrapperSize.value
        } else {
          offset = _offset
        }
        setNavOffset(offset)
      }
    }

    const judgePreNextStatus = () => {
      preReached.value = navOffset.value === 0
      nextReached.value = navSize.value - navOffset.value <= navWrapperSize.value
    }

    const update = () => {
      setNavElSize()
      setNavPreNextElSize()
      setSelectedElSize()
      setSelectedElOffset()
      updateNavBarStyle()
      judgePreNextStatus()
    }

    watch(
      navOffset,
      val => {
        if (navElRef.value) {
          navElRef.value.style.transform = `translate${isHorizontal.value ? 'X' : 'Y'}(-${val}px)`
          judgePreNextStatus()
          updateNavBarStyle()
        }
      },
      {
        flush: 'post',
      },
    )

    let isAddTabs = false

    watch(
      () => props.tabs,
      (val = [], oldVal = []) => {
        update()
        isAddTabs = val.length > oldVal.length
      },
      {
        flush: 'post',
      },
    )

    watch(
      navSize,
      (currentSize, oldSize) => {
        let offset = navOffset.value
        if (currentSize < oldSize && !isAddTabs) {
          offset += currentSize - oldSize
          setNavOffset(offset > 0 ? offset : 0)
        }
      },
      {
        flush: 'post',
      },
    )

    watch(selectedKey, val => {
      const hasSelectedKey = props.tabs?.find(item => {
        return val === item.key
      })
      if (!hasSelectedKey) {
        selectedElRef.value = null
      }
    })

    watch(
      selectedElRef,
      () => {
        setSelectedElSize()
        setSelectedElOffset()
        setSelectedElOffset()
        updateSelectedOffset()
        updateNavBarStyle()
      },
      {
        flush: 'post',
      },
    )

    useResizeObserver(navWrapperElRef, update)

    provide(tabsToken, {
      selectedKey,
      selectedElRef,
      mergedPrefixCls,
      handleTabClick,
    })

    return () => {
      let defaultSelectedKey: VKey = 1

      const tabVNodes =
        props.tabs?.map((item, index) => {
          if (isNil(item.key)) {
            item.key = index + 1
          } else if (index === 0) {
            defaultSelectedKey = item.key as VKey
          }
          return item
        }) ?? []

      return (
        <div class={classes.value}>
          <div class={navWrapperClass.value} ref={navWrapperElRef}>
            {hasScroll.value && (
              <IxIcon
                class={navPreClasses.value}
                name={isHorizontal.value ? 'left' : 'up'}
                onClick={handlePreClick}
                ref={navPreElRef}
              />
            )}
            <div class={`${mergedPrefixCls.value}-nav`} ref={navElRef}>
              {
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                tabVNodes.map((vnode: any) => (
                  <TabNav
                    {...vnode.props}
                    key={vnode.key}
                    defaultSelectedKey={defaultSelectedKey}
                    v-slots={{ title: vnode.children?.title }}
                  />
                ))
              }
            </div>
            {hasScroll.value && (
              <IxIcon
                class={navNextClasses.value}
                name={isHorizontal.value ? 'right' : 'down'}
                onClick={handleNextClick}
              />
            )}
            {!isSegmentType.value && <div class={`${mergedPrefixCls.value}-nav-border`}></div>}
            {isLineType.value && (
              <div class={`${mergedPrefixCls.value}-nav-bar`} ref={navBarElRef} style={barStyle.value}></div>
            )}
          </div>
          <div class={`${mergedPrefixCls.value}-pane-wrapper`}>
            {filterTabVNodes(props, tabVNodes, selectedKey, defaultSelectedKey)}
          </div>
        </div>
      )
    }
  },
})
