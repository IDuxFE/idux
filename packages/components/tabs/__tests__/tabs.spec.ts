import { MountingOptions, mount } from '@vue/test-utils'
import { h } from 'vue'

import { renderWork } from '@tests'

import Tab from '../src/Tab'
import Tabs from '../src/Tabs'
import { TabsProps } from '../src/types'

const defaultSlots = [
  h(Tab, { key: 'tab1' }, { default: () => '内容1' }),
  h(Tab, { key: 'tab2' }, { default: () => '内容2' }),
  h(Tab, { key: 'tab3' }, { default: () => '内容3' }),
]

describe('Tabs', () => {
  const TabsMount = (options?: MountingOptions<Partial<TabsProps>>) => {
    const { slots, ...rest } = options || {}
    const mergedOptions = {
      slots: { default: () => defaultSlots, ...slots },
      ...rest,
    } as MountingOptions<TabsProps>
    return mount(Tabs, mergedOptions)
  }

  renderWork<TabsProps>(Tabs)

  test('v-model:selectedKey work', async () => {
    const onUpdateSelectedKey = jest.fn()
    const onTabClick = jest.fn()
    const wrapper = TabsMount({
      props: {
        selectedKey: 'tab1',
        'onUpdate:selectedKey': onUpdateSelectedKey,
        onTabClick,
      },
    })

    expect(wrapper.findAll('.ix-tabs-nav-tab')[0].classes()).toContain('ix-tabs-nav-tab-selected')

    await wrapper.setProps({ selectedKey: 'tab2' })

    expect(wrapper.findAll('.ix-tabs-nav-tab')[0].classes()).not.toContain('ix-tabs-nav-tab-selected')

    await wrapper.find('.ix-tabs-nav-tab').trigger('click')

    expect(onUpdateSelectedKey).toBeCalledTimes(1)
    expect(onTabClick).toBeCalledTimes(1)
  })

  test('type work', async () => {
    const wrapper = TabsMount()

    expect(wrapper.classes()).toContain('ix-tabs-card')

    await wrapper.setProps({ type: 'line' })

    expect(wrapper.classes()).toContain('ix-tabs-line')

    await wrapper.setProps({ type: 'segment' })

    expect(wrapper.classes()).toContain('ix-tabs-segment')
  })

  test('mode work', async () => {
    const wrapper = TabsMount({
      props: {
        type: 'segment',
        mode: 'primary',
      },
    })

    expect(wrapper.classes()).toContain('ix-tabs-segment')
    expect(wrapper.classes()).toContain('ix-tabs-nav-primary')

    await wrapper.setProps({ type: 'card' })

    expect(wrapper.classes()).not.toContain('ix-tabs-nav-primary')
  })

  test('forceRender work', async () => {
    const wrapper = TabsMount({
      props: {
        selectedKey: 'tab1',
      },
    })

    expect(wrapper.findAll('.ix-tabs-pane').length).toBe(1)

    await wrapper.setProps({ forceRender: true })

    expect(wrapper.findAll('.ix-tabs-pane').length).toBe(3)
  })

  test('placement work', async () => {
    const wrapper = TabsMount({
      props: {
        type: 'line',
        placement: 'start',
      },
    })

    expect(wrapper.classes()).toContain('ix-tabs-nav-start')

    await wrapper.setProps({ type: 'card' })

    expect(wrapper.classes()).not.toContain('ix-tabs-nav-start')
  })

  test('only accept tab children work', async () => {
    const wrapper = TabsMount({
      props: {
        forceRender: true,
      },
      slots: {
        default: () => [h(Tab, {}, { title: () => '内容1' }), h('div', {}, { default: () => '内容2' })],
      },
    })

    expect(wrapper.findAll('.ix-tabs-nav-tab').length).toBe(1)
    expect(wrapper.findAll('.ix-tabs-pane').length).toBe(1)
  })

  describe('Tab', () => {
    test('title work', async () => {
      const wrapper = TabsMount({
        slots: {
          default: () => h(Tab, {}, { title: () => '内容1' }),
        },
      })

      expect(wrapper.find('.ix-tabs-nav-tab').text()).toBe('内容1')
    })

    test('disabled work', async () => {
      const wrapper = TabsMount({
        slots: {
          default: () => h(Tab, { disabled: true }, { tab: () => '内容1' }),
        },
      })

      expect(wrapper.find('.ix-tabs-nav-tab').classes()).toContain('ix-tabs-nav-tab-disabled')
    })

    test('forceRender work', async () => {
      const wrapper = TabsMount({
        props: {
          selectedKey: 'tab1',
        },
        slots: {
          default: () => [
            h(Tab, { key: 'tab1', forceRender: true }, { tab: () => '内容1' }),
            h(Tab, { key: 'tab2' }, { tab: () => '内容2' }),
          ],
        },
      })

      expect(wrapper.findAll('.ix-tabs-pane').length).toBe(1)

      await wrapper.setProps({ selectedKey: 'tab2' })

      expect(wrapper.findAll('.ix-tabs-pane').length).toBe(2)
    })
  })
})
