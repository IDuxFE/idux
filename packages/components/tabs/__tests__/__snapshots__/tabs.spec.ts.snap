// Vitest Snapshot v1

exports[`Tabs > render work 1`] = `
"<div class=\\"ix-tabs ix-tabs-md ix-tabs-card ix-tabs-nav-top\\">
  <div class=\\"ix-tabs-nav-wrapper\\">
    <div class=\\"ix-tabs-nav\\">
      <div style=\\"transform: translateX(-0px);\\" class=\\"ix-tabs-nav-list\\">
        <!---->
      </div>
    </div>
    <div class=\\"ix-tabs-nav-operations ix-tabs-nav-operations-hidden\\"><button class=\\"ix-button ix-button-icon-only ix-button-text ix-button-square ix-button-md\\" type=\\"button\\"><i class=\\"ix-icon ix-icon-more\\" role=\\"img\\" aria-label=\\"more\\"></i>
        <!---->
      </button>
      <!---->
      <!---->
    </div>
    <div class=\\"ix-tabs-nav-border\\"></div>
    <!---->
  </div>
  <div class=\\"ix-tabs-pane-wrapper\\"></div>
</div>"
`;
